server/py/framework/db/sqldb/db.py

```diff
@@ -331,6 +331,42 @@ def update_run(self, session, updates: dict, uid: str, project: str, iter: int =
         self._delete_empty_labels(session, Run.Label)
         return run.struct
 
+    def set_run_retrying_state(
+        self,
+        session: Session,
+        project: str,
+        uid: str,
+        retrying: bool,
+    ) -> dict:
+        """
+        Atomically LOCK the run row FOR UPDATE, then either add or remove the
+        `retrying` label according to the `retrying` argument.
+        """
+        run = self._get_run(session, uid, project, iteration=0, with_for_update=True)
+        if not run:
+            raise mlrun.errors.MLRunNotFoundError(f"Run {project}/{uid} not found")
+
+        struct = run.struct
+        labels = run_labels(struct)
+
+        if retrying:
+            if mlrun_constants.MLRunInternalLabels.retrying in labels:
+                # flush and commit so the lock is released immediately
+                session.commit()
+                raise mlrun.errors.MLRunConflictError
+            else:
+                labels[mlrun_constants.MLRunInternalLabels.retrying] = "true"
+                # optionally bump counter here in follow-up
+                # labels["rerun_counter"] = str(int(labels.get("rerun_counter","0")) + 1)
+        else:
+            labels.pop("retrying", None)
```

Comment:
Not a completely necessary optimization, however, it does indeed make the decision tree shorter (exit fast) and makes the code nicer with less indentations:
```suggestion
        if not retrying:
            labels.pop("retrying", None)
        elif mlrun_constants.MLRunInternalLabels.retrying in labels:
            # flush and commit so the lock is released immediately
            session.commit()
            raise mlrun.errors.MLRunConflictError
        else:
            labels[mlrun_constants.MLRunInternalLabels.retrying] = "true"
```